Add two level transfer operator between non-nested levels

fdrmrc opened the pull request:
The present PR, written with @peterrum, introduces a two-level transfer operator between (possibly) non-nested multigrid levels, called `MGTwoLevelTransferNonNested`. Since many of the tasks are similar to the ones already implemented in `MGTwoLevelTransfer`, a base class `MGTwoLevelTransferBase` has been added. The current implementation assumes continuous and scalar elements.

This transfer is supposed to be used exactly in the same way as the classical `MGTwoLevelTransfer`, as shown in some of the tests, where levels are created out of randomly distorted and refined triangulations. Other strategies may be chosen, as long as each level can described by a deal.II tria.

If it helps the review process, this PR can be split into smaller PRs.

luca-heltai commented on include/deal.II/multigrid/mg_transfer_global_coarsening.h:
```diff
@@ -153,26 +156,222 @@ namespace MGTransferGlobalCoarseningTools
 
 
 
```
Can you add a documentation to the class, explaining what the class is doing?

luca-heltai commented on include/deal.II/base/mpi_remote_point_evaluation.h:
```diff
@@ -108,6 +108,16 @@ namespace Utilities
         std::vector<Point<dim>> reference_point_values;
       };
 
```
Can you move the implementation to the cc file?

luca-heltai replied:
Also, this seems a small, self contained change. Can you make a separate PR for this one?

luca-heltai commented on include/deal.II/non_matching/mapping_info.h:
```diff
@@ -96,10 +96,10 @@ namespace NonMatching
      * with the "unfiltered" index. The default argument for
      * @p n_unfiltered_cells disables this built-in compression.
      */
```
All changes to this file could go in a separate PR.

luca-heltai commented on tests/multigrid-global-coarsening/non_nested_multigrid_01.cc:
```diff
@@ -0,0 +1,155 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 - 2022 by the deal.II authors
```
Here and below. 2023

luca-heltai reviewed (changes requested):
Very nice! I'm not a great expert on the topic, so I'll let someone else comment on this. Many of the small changes can go in separate PRs, and you should squash the changes to two or three commits to make them readale.

peterrum commented:
@fdrmrc The 3 PRs are merged. Could you update this PR?

peterrum commented on include/deal.II/non_matching/mapping_info.h:
```diff
@@ -471,7 +471,7 @@ namespace NonMatching
 
   template <int dim, int spacedim>
-  template <typename Iterator>
+  template <typename Iterators>
```
```suggestion
```

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+        constraint_info.read_write_operation(
+          reader,
+          src,
+          reinterpret_cast<VectorizedArrayType *>(solution_values.data()),
```
Don't we support plain (non-vectorized) read access in `ConstraintInfo`? I guess this relates to the newly opened issue #15162.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+  if (rpe.is_map_unique() == false)
+    {
+      const auto evaluation_point_results_temp = evaluation_point_results;
+      evaluation_point_results.assign(rpe.get_point_ptrs().size() - 1, 0);
```
Let us avoid touching data twice. First do
```suggestion
      evaluation_point_results.clear();
      evaluation_point_results.reserve(rpe.get_point_ptrs().size() - 1);
```

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+            result += evaluation_point_results_temp[ptr[i] + j];
+          result /= n_entries;
+
+          evaluation_point_results[i] = result;
```
and then do
```suggestion
          evaluation_point_results.push_back(result);
```

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+
+      for (unsigned int i = 0; i < ptr.size() - 1; ++i)
+        {
+          const auto n_entries = ptr[i + 1] - ptr[i];
+          if (n_entries == 0)
+            continue;
+
+          evaluation_point_results[i] /= n_entries;
```
Is this really the transpose of the function above? This would mean that the `process_and_evaluate` function works asymmetrically, because in case `n_entries` is larger than 1 somewhere, the lengths of `evaluation_points_results` here is different from the `evaluation_points_results` used in `prolongate_and_add`. Note that the index used here is `i`, whereas it is `j` (aka `ptr[i] + ...`) above. I am confused.

peterrum replied:
Let me answer this one. Indeed the implementation in `RPE` not completely symmetric, which is related to the application cases for which each function was written.

The function `evaluate_and_process()` collect for each point the contributions from all cells.

However, `process_and_evaluate()` sends the data at the point to the relevant cells. The same data is sent to all cells. The copying is done here:

https://github.com/dealii/dealii/blob/09cb1410a77d398fd711cbcf0f9cb07e960c323b/include/deal.II/base/mpi_remote_point_evaluation.h#L467-L475

`evaluation_point_results[i] /= n_entries;` followed by the copying is as far as I see the transpose of averaging, which is done during prolongation.

kronbichler commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+
+template <int dim, typename Number>
+void
+MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
+  enable_inplace_operations_if_possible(
+    const std::shared_ptr<const Utilities::MPI::Partitioner>
+      &external_partitioner_coarse,
+    const std::shared_ptr<const Utilities::MPI::Partitioner>
+      &external_partitioner_fine)
+{
+  this->internal_enable_inplace_operations_if_possible(
+    external_partitioner_coarse, external_partitioner_fine, constraint_info);
```
Related to my comment above, this at least indicates that `enable_inplace_operations_if_possible` in `MGTwoLevelTransferBase` should not be pure virtual and directly expose this functionality. Otherwise, see the other comments.

kronbichler commented on tests/multigrid-global-coarsening/mg_transfer_util.h:
```diff
@@ -167,4 +172,138 @@ test_transfer_operator(
+    constraint_fine.distribute(dst);
+
+    // Visualize the result
+    {
+      DataOut<2> data_out;
+      data_out.attach_dof_handler(dof_handler_coarse);
+      data_out.add_data_vector(
+        src,
+        "coarse_solution",
+        DataOut_DoFData<dim, dim>::DataVectorType::type_dof_data);
+      data_out.build_patches();
+      std::ofstream output_coarse("coarse_sol.vtk");
+      data_out.write_vtk(output_coarse);
+      data_out.clear();
+      std::ofstream output_fine("prolonged_sol.vtk");
+      data_out.attach_dof_handler(dof_handler_fine);
+      data_out.add_data_vector(
+        dst,
+        "prolonged_solution",
+        DataOut_DoFData<dim, dim>::DataVectorType::type_dof_data);
+
+      data_out.build_patches();
+      data_out.write_vtk(output_fine);
+    }
```
Do we really want to write the result? Should we check it? As tests are run on systems with limited disk performance, I would prefer not writing the data.

kronbichler reviewed (commented):
Nice, this is great work, thanks a lot @fdrmrc (and @peterrum). I particularly like the extensive tests. There are some things that I would like to have fixed, plus some other parts where I added comments to be addressed later.

luca-heltai commented:
Depends on #15167.

peterrum reviewed (commented):
Don't forget to add a change-log entry.

peterrum commented:
/rebuild

peterrum commented on include/deal.II/matrix_free/tensor_product_kernels.h:
```diff
@@ -3482,10 +3482,10 @@ namespace internal
-        for (unsigned int i = 0; i < array_size; ++i)
+        for (unsigned int j = 0; j < array_size; ++j)
           {
-            shape_values_x[i] = shapes[i][0][0];
-            shape_derivs_x[i] = shapes[i][1][0];
+            shape_values_x[j] = shapes[j][0][0];
+            shape_derivs_x[j] = shapes[j][1][0];
```
@kronbichler @bergbauer We needed to rename `i` to some other name, since the compiler was warning that there are two `i` shadowing each other.

fdrmrc replied:
Here's the error message: https://github.com/dealii/dealii/actions/runs/4884140277/jobs/8716458324#step:10:2319

bergbauer replied:
Odd that it only complains about this now but yes this should be a different variable.

fdrmrc commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3175,6 +3176,58 @@ MGTwoLevelTransferBase<LinearAlgebra::distributed::Vector<Number>>::
+      return embedded_partitioner;
+    }
+  } // namespace
+} // namespace internal
```
Fixed, thanks.

fdrmrc commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+          typename Triangulation<dim>::active_cell_iterator cell = {
+            &rpe.get_triangulation(),
+            cell_data.cells[i].first,
+            cell_data.cells[i].second};
```
Fixed.

fdrmrc commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+  std::vector<types::global_dof_index>                        dof_indices(
+    dof_handler_fine.get_fe().n_dofs_per_cell());
+
+  const auto &local_indices_fine = dof_handler_fine.locally_owned_dofs();
+
+  Quadrature<dim> quadrature(unit_pts);
+  FEValues<dim>   fe_values(mapping_fine,
+                          fe_space,
+                          quadrature,
+                          update_quadrature_points);
+
+  for (const auto &cell : dof_handler_fine.active_cell_iterators() |
+                            IteratorFilters::LocallyOwnedCell())
+    {
+      fe_values.reinit(cell);
+      cell->get_dof_indices(dof_indices);
+
+      for (unsigned int i = 0; i < dof_indices.size(); ++i)
+        if (local_indices_fine.is_element(dof_indices[i]) &&
+            (constraint_fine.is_constrained(dof_indices[i]) == false))
+          points_all.emplace_back(local_indices_fine.index_within_set(
+                                    dof_indices[i]),
+                                  fe_values.quadrature_point(i));
+    }
+
+  std::sort(points_all.begin(),
+            points_all.end(),
+            [](const auto &a, const auto &b) { return a.first < b.first; });
+  points_all.erase(std::unique(points_all.begin(),
+                               points_all.end(),
+                               [](const auto &a, const auto &b) {
+                                 return a.first == b.first;
+                               }),
+                   points_all.end());
+
+  for (const auto &i : points_all)
+    {
+      this->level_dof_indices_fine.push_back(i.first);
+      points.push_back(i.second);
+    }
```
It is, modulo a post-processing to extract locally owned dofs. However, this doesn't generalize to the case of multiple components.

fdrmrc commented:
I addressed some of the comments in the review. As we discussed in person last week, the choice to have `enable_inplace_operations_if_possible` directly in the base class was done in order to have a unified interface.

fdrmrc commented on include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:
```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+                                         unit_points.end());
+
+          cell_iterators.emplace_back(cell);
+          unit_points_vector.emplace_back(unit_points_vector_cell);
```
Yes, thanks.

kronbichler reviewed (approved):
Thank you for the adjustments, and thanks for working on the feature, this is really cool. I was involved in advising @fdrmrc so I would appreciate an additional look at this before merging.

peterrum commented:
@luca-heltai We are happy with the state. There are some follow-up tasks we have collected in https://github.com/dealii/dealii/issues/15148. Could you also take another look and see if you are happy now!?
